nixos/modules/virtualisation/xen-dom0.nix

```diff
@@ -1,213 +1,925 @@
+        message = "Xen's debugging features are enabled, but logging is disabled. This is most likely not what you want.";
+      }
+      {
+        assertion = cfg.store.settings.quota.maxWatchEvents >= cfg.store.settings.quota.maxOutstanding;
+        message = ''
+          Upstream Xen recommends that maxWatchEvents be equal to or greater than maxOutstanding,
+          in order to mitigate denial of service attacks from malicious frontends.
+        '';
+      }
+    ];
 
+    environment.systemPackages = [
+      cfg.package
+      cfg.qemu.package
+    ];
 
-  ###### implementation
+    virtualisation.xen.bootParams =
+      lib.lists.optionals cfg.trace [
+        "loglvl=all"
+        "guest_loglvl=all"
+      ]
+      ++
+        lib.lists.optional (cfg.dom0Resources.memory != 0)
+          "dom0_mem=${toString cfg.dom0Resources.memory}M${
+            lib.strings.optionalString (
+              cfg.dom0Resources.memory != cfg.dom0Resources.maxMemory
+            ) ",max:${toString cfg.dom0Resources.maxMemory}M"
+          }"
+      ++ lib.lists.optional (
+        cfg.dom0Resources.maxVCPUs != 0
+      ) "dom0_max_vcpus=${toString cfg.dom0Resources.maxVCPUs}";
 
-  config = mkIf cfg.enable {
-    assertions = [ {
-      assertion = pkgs.stdenv.isx86_64;
-      message = "Xen currently not supported on ${pkgs.stdenv.hostPlatform.system}";
-    } {
-      assertion = config.boot.loader.grub.enable && (config.boot.loader.grub.efiSupport == false);
-      message = "Xen currently does not support EFI boot";
-    } ];
-
-    virtualisation.xen.package = mkDefault pkgs.xen;
-    virtualisation.xen.package-qemu = mkDefault pkgs.xen;
-    virtualisation.xen.stored = mkDefault "${cfg.package}/bin/oxenstored";
-
-    environment.systemPackages = [ cfg.package ];
-
-    boot.kernelModules =
-      [ "xen-evtchn" "xen-gntdev" "xen-gntalloc" "xen-blkback" "xen-netback"
-        "xen-pciback" "evtchn" "gntdev" "netbk" "blkbk" "xen-scsibk"
-        "usbbk" "pciback" "xen-acpi-processor" "blktap2" "tun" "netxen_nic"
-        "xen_wdt" "xen-acpi-processor" "xen-privcmd" "xen-scsiback"
+    boot = {
+      kernelModules = [
+        "xen-evtchn"
+        "xen-gntdev"
+        "xen-gntalloc"
+        "xen-blkback"
+        "xen-netback"
+        "xen-pciback"
+        "evtchn"
+        "gntdev"
+        "netbk"
+        "blkbk"
+        "xen-scsibk"
+        "usbbk"
+        "pciback"
+        "xen-acpi-processor"
+        "blktap2"
+        "tun"
+        "netxen_nic"
+        "xen_wdt"
+        "xen-acpi-processor"
+        "xen-privcmd"
+        "xen-scsiback"
         "xenfs"
       ];
 
-    # The xenfs module is needed in system.activationScripts.xen, but
-    # the modprobe command there fails silently. Include xenfs in the
-    # initrd as a work around.
-    boot.initrd.kernelModules = [ "xenfs" ];
+      # The xenfs module is needed in system.activationScripts.xenfs.
+      initrd.kernelModules = [ "xenfs" ];
```

Comment:
There is an option `boot.initrd.includeDefaultModules` if setting that to `false` would break things then we should make a warning about it.

Reply:
We don't set `includeDefaultModules` to false though? It also shouldn't break anything as we load `xenfs` explicitly.

Reply:
The idea is more reliability so that when a user misconfigures the kernel modules to get a warning about it instead of figuring it out on deployment.

Reply:
I don't see how a user can misconfigure the kernel modules, except by explicitly overriding the module and removing the Xen modules.